### PR TITLE
fix jit trace error for model forward sequence is not aligned with jit.trace tuple input sequence, update related doc

### PR DESCRIPTION
**Pull request by sywangyi**
Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

@sgugger

### CODE REVIEW

**Comment by sywangyi** [comment]
@liangan1 @jianan-gu @yao-matrix please help review

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1253,18 +1253,35 @@ def torch_jit_model_eval(self, model, dataloader, training=False):
                 return model
             jit_inputs = []
             example_batch = next(iter(dataloader))
+            if "labels" in example_batch:
```
Comment (liangan1) [inline src/transformers/trainer.py:1256]: This change doesn't make sense. there is no relationship between label, label_smotherer and jit.trace.

Comment (sywangyi) [reply src/transformers/trainer.py:1256]: this logic is copied from trainer.py. if label smoother option is opened. label should be pop from input.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1253,18 +1253,36 @@ def torch_jit_model_eval(self, model, dataloader, training=False):
...
+
             try:
                 jit_model = model.eval()
-                with ContextManagers([self.autocast_smart_context_manager(), torch.no_grad()]):
+                with ContextManagers(
+                    [
+                        torch.cpu.amp.autocast(cache_enabled=False, dtype=self.amp_dtype)
+                        if self.use_cpu_amp
+                        else contextlib.nullcontext(),
+                        torch.no_grad(),
```
Comment (sgugger) [inline src/transformers/trainer.py:1277]: Why are you removing support for GPU mixed precision here?

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1253,18 +1253,36 @@ def torch_jit_model_eval(self, model, dataloader, training=False):
                 return model
             jit_inputs = []
             example_batch = next(iter(dataloader))
+            if "labels" in example_batch:
+                if self.label_smoother is not None:
+                    example_batch.pop("labels")
+                elif version.parse(version.parse(torch.__version__).base_version) < version.parse("1.14.0"):
```
Comment (sgugger) [inline src/transformers/trainer.py:1259]: This mean the whole eval will fail in 99.99% of the case when users do not use label smoothing (it's a training technique)

**Review by sgugger — CHANGES_REQUESTED** [review]
This PR breaks the JIT eval in pretty much all cases unless on PyTorch nighly, unless label smoothing is activated.
Label smoothing is a training technique, so it's very unlikely a user will combine a jit eval with label smoothing (when jsut running evaluation). This PR therefore can't be accepted as it is.

File: src/transformers/trainer.py (hunk 2 above)
Comment (sywangyi) [reply src/transformers/trainer.py]: thanks for the reminding, add it back. and verified in GPU

File: src/transformers/trainer.py (hunk 3 above)
Comment (sywangyi) [reply src/transformers/trainer.py]: the fix of jit trace failure is https://github.com/pytorch/pytorch/pull/81623, I add it to the doc to let user know it, which is expected to be merged in pytorch1.14.0. However, we will catch the trace failure issue and fallback to the original path

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -1253,20 +1253,34 @@ def torch_jit_model_eval(self, model, dataloader, training=False):
...
+            if self.label_smoother is not None and "labels" in example_batch:
+                example_batch.pop("labels")
             for key in example_batch:
                 example_tensor = torch.ones_like(example_batch[key])
                 jit_inputs.append(example_tensor)
             jit_inputs = tuple(jit_inputs)
             try:
                 jit_model = model.eval()
-                with ContextManagers([self.autocast_smart_context_manager(), torch.no_grad()]):
+                with ContextManagers(
```
Comment (sgugger) [inline src/transformers/trainer.py:1265]: This is the logic that is in `self.autocast_smart_context_manager()` if I'm not mistaken so let's no change anything here.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -1717,7 +1731,6 @@ def _inner_training_loop(
 
             step = -1
             for step, inputs in enumerate(epoch_iterator):
-
```
Comment (sgugger) [inline src/transformers/trainer.py:1720]: Please let's not add unrelated changes like all those empty lines removed.

**Review by sgugger — COMMENTED** [review]
I still do not understand the main problem: it looks like JIT does not support dictionary inputs which are used in every model in Transformers. Classification models are not the only ones using the `labels` key, all task-specific models do... and a model that a user wants to evaluate will very likely have a dataset with labels. The proposed workaround to use label smoothing makes no sense for an evaluation.

It looks like this integration has maybe be merged too quickly and doesn't actually work or are there models that can be evaluated with it?

File: src/transformers/trainer.py (hunk 4 above)
Comment (sywangyi) [reply src/transformers/trainer.py:1265]: Hi, sgugger, the key point is we need set "cache_enabled=False" here, self.autocast_smart_context_manager() do not have suchargument setting. or else, torch+bf16/fp16+jit.trace will fail here.

Comment (sywangyi) [reply src/transformers/trainer.py:1265]: Hi, sgugger, the key point is we need set "cache_enabled=False" here, self.autocast_smart_context_manager() do not have such argument setting. or else, torch+bf16/fp16+jit.trace will fail here.

**Comment by sywangyi** [comment]
> I still do not understand the main problem: it looks like JIT does not support dictionary inputs which are used in every model in Transformers. Classification models are not the only ones using the `labels` key, all task-specific models do... and a model that a user wants to evaluate will very likely have a dataset with labels. The proposed workaround to use label smoothing makes no sense for an evaluation.
> 
> It looks like this integration has maybe be merged too quickly and doesn't actually work or are there models that can be evaluated with it?

yes. all the cases containing "labels" will fail in jit.trace, while other case like QnA could pass. it's pytorch limitation for jit.trace which only support tuple input now, Intel has commited a PR(https://github.com/pytorch/pytorch/pull/81623) for this and expected to be released in pytorch 1.14 (I also added it in doc).

It we would like to jit.trace successfully for such case, the other option is to modify the model like below, making forward input sequence like tuple input sequence..., 
--- a/src/transformers/models/distilbert/modeling_distilbert.py
+++ b/src/transformers/models/distilbert/modeling_distilbert.py
@@ -731,11 +731,11 @@ class DistilBertForSequenceClassification(DistilBertPreTrainedModel):
     )
     def forward(
         self,
+        labels: Optional[torch.LongTensor] = None,
         input_ids: Optional[torch.Tensor] = None,
         attention_mask: Optional[torch.Tensor] = None,
         head_mask: Optional[torch.Tensor] = None,
         inputs_embeds: Optional[torch.Tensor] = None,
-        labels: Optional[torch.LongTensor] = None,
         output_attentions: Optional[bool] = None,
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,

File: src/transformers/trainer.py (hunk 5 above)
Comment (sywangyi) [reply src/transformers/trainer.py]: done

**Comment by jianan-gu** [comment]
The key point of the jit error cases met here is that jit cannot well handle the case that the dictionary forward parameter order does not match the dataset input order, not specific to whether there are "labels" or not. And  to improve PyTorch jit ability to solve this issue, we landed  https://github.com/pytorch/pytorch/pull/81623 in PyTorch;

For the usage of model inference with jit, for now, there could be many cases that natively get the benefits, like models running question-answering example mentioned above;

For these failed model inferences with jit cases, we are capturing this with the exception here to make it fallback and use logging to notify users; Meanwhile, these failed cases shall work when PyTorch contains the [PR](https://github.com/pytorch/pytorch/pull/81623), (expect in next release);

Besides, bringing "label smoothing" here with jit is not that reasonable since it would be confusing for users.

**Comment by sywangyi** [comment]
Hi, @sgugger to make it a clear, I file a issue to record the issue I meet https://github.com/huggingface/transformers/issues/19973. also I agree that "label smoothing" is a training skill and I have removed it in inference part.

File: src/transformers/trainer.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/trainer.py:1263]: Can we add that argument to the method then?

**Review by sgugger — COMMENTED** [review]
You are a bit beating around the bush here: are there any models with a head where this feature can be used right now without hacks? I understand support in PyTorch is coming in the next version for dictionaries, but I think this feature was just added to early. Can the doc explicitly mention that the feature requires a nightly install?

**Comment by sywangyi** [comment]
> You are a bit beating around the bush here: are there any models with a head where this feature can be used right now without hacks? I understand support in PyTorch is coming in the next version for dictionaries, but I think this feature was just added to early. Can the doc explicitly mention that the feature requires a nightly install?

Hi, sgugger 
for pytorch >= 1.14.0 (nightly version is 1.14.0). jit could benefit any models for predict and eval.
for pytorch < 1.14.0.  jit could benefit models like "Question and Answer"， whose forward parameter order matches the tuple input order in jit.trace. If we meet case like "text classification"，whose forward parameter order does not matches the tuple input order in jit.trace in evaluation, jit trace will fail and we are capturing this with the exception here to make it fallback and use logging to notify users

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -1251,20 +1251,34 @@ def torch_jit_model_eval(self, model, dataloader, training=False):
...
+                with ContextManagers([self.autocast_smart_context_manager(cache_enabled=False), torch.no_grad()]):
+                    if version.parse(version.parse(torch.__version__).base_version) >= version.parse("1.14.0"):
+                        if isinstance(example_batch, dict):
+                            jit_model = torch.jit.trace(jit_model, example_kwarg_inputs=example_batch, strict=False)
+                        else:
+                            jit_model = torch.jit.trace(
+                                jit_model,
+                                example_kwarg_inputs={key: example_batch[key] for key in example_batch},
+                                strict=False,
+                            )
```
Comment (sgugger) [inline src/transformers/trainer.py:1267]: I don't fully understand this path here, as `example_batch` is not a dict here, but you are using it as one.

**Review by sgugger — COMMENTED** [review]
Thanks for the precision. Could you add all of this to the documentation?
Also have one last question on the actual code.

File: src/transformers/trainer.py (hunk 6 above)
Comment (sywangyi) [reply src/transformers/trainer.py:1267]: I just meet the example_batch is a class like transformers.feature_extraction_utils.BatchFeature in audio-classification example. because the BatchFeature implements the API like __getitem__(), I could operate it like a dict

**Comment by sywangyi** [comment]
> Thanks for the precision. Could you add all of this to the documentation? Also have one last question on the actual code.

which document would you recommend to add this, since it's not cpu specific.

File: src/transformers/trainer.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/trainer.py:1267]: Understood!

**Comment by sgugger** [comment]
> which document would you recommend to add this

Every time the jit eval is mentioned.

File: docs/source/en/perf_infer_cpu.mdx [hunk 7]
```diff
@@ -29,7 +29,16 @@ Check more detailed information for [IPEX Graph Optimization](https://intel.gith
...
 
 ### Usage of JIT-mode
-To enable jit mode in Trainer, users should add `jit_mode_eval` in Trainer command arguments.
+To enable JIT-mode in Trainer for evaluaion or prediction, users should add `jit_mode_eval` in Trainer command arguments.
+
+<Tip warning={true}>
+
+for PyTorch >= 1.14.0. JIT-mode could benefit any models for prediction and evaluaion since dict input is supported in jit.trace
+
+for PyTorch < 1.14.0. JIT-mode could benefit models like "Question and Answer", whose forward parameter order matches the tuple input order in jit.trace.
```
Comment (sgugger) [inline docs/source/en/perf_infer_cpu.mdx:38]:
```suggestion
for PyTorch < 1.14.0. JIT-mode could benefit modelswhose forward parameter order matches the tuple input order in jit.trace, like question-answering models.
```

**Review by sgugger — COMMENTED** [review]
Thanks for adding the documentation. Just left a few last nits.

File: docs/source/en/perf_infer_cpu.mdx [hunk 8]
```diff
@@ -29,7 +29,16 @@ Check more detailed information for [IPEX Graph Optimization](https://intel.gith
...
+If we meet case like "text classification", whose forward parameter order does not match the tuple input order in jit.trace, jit.trace will fail and we are capturing this with the exception here to make it fallback and logging is used to notify users.
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply docs/source/en/perf_infer_cpu.mdx]: done

**Review by sgugger — APPROVED** [review]
Thanks for iterating on this.
